File: lib/product/models/payload_action.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'payload_action.freezed.dart';
part 'payload_action.g.dart';

@freezed
sealed class PayloadAction with _$PayloadAction {
  const factory PayloadAction.wait({required int duration}) = WaitAction;
  const factory PayloadAction.swipe({required String direction}) = SwipeAction;
  const factory PayloadAction.click({required String nodeText}) = ClickAction;
  const factory PayloadAction.launch({required String packageName}) =
      LaunchAction;
  const factory PayloadAction.typeText({required String text}) =
      TypeTextAction;

  factory PayloadAction.fromJson(Map<String, dynamic> json) =>
      _$PayloadActionFromJson(json);
}

File: android/app/src/main/kotlin/com/ghosttraffic/ghost_traffic_lab/models/PayloadAction.kt
package com.ghosttraffic.ghost_traffic_lab.models

import org.json.JSONArray
import org.json.JSONObject

sealed class PayloadAction {
    data class Wait(val duration: Long) : PayloadAction()
    data class Swipe(val direction: String) : PayloadAction()
    data class Click(val nodeText: String) : PayloadAction()
    data class Launch(val packageName: String) : PayloadAction()
    data class TypeText(val text: String) : PayloadAction()

    companion object {
        fun fromJsonArray(json: String): List<PayloadAction> {
            val array = JSONArray(json)
            return (0 until array.length()).map { i ->
                fromJsonObject(array.getJSONObject(i))
            }
        }

        private fun fromJsonObject(obj: JSONObject): PayloadAction {
            return when (obj.getString("action")) {
                "wait" -> Wait(obj.getLong("duration"))
                "swipe" -> Swipe(obj.getString("direction"))
                "click" -> Click(obj.getString("node_text"))
                "launch" -> Launch(obj.getString("package_name"))
                "type" -> TypeText(obj.getString("text"))
                else -> throw IllegalArgumentException(
                    "Unknown action: ${obj.getString("action")}"
                )
            }
        }
    }
}
